client/me/purchases/manage-purchase/index.jsx

```diff
@@ -988,6 +1017,7 @@ function PurchasesQueryComponent( { isSiteLevel, selectedSiteId } ) {
 export default connect(
 	( state, props ) => {
 		const purchase = getByPurchaseId( state, props.purchaseId );
+		const difmTieredPurchaseDetails = getDIFMTieredPurchaseDetails( state, props.purchaseId );
```

Comment:
Actually I was modelling it as a selector so it can be used anywhere that state slice needs to be returned. I actually don't understand why it needs to go in a separate lib directory when this is clearly a calculation of existing state 🤔.

Reply:
The calculation is on the `purchase` object, which could come from different places - it could be from the global redux state (like here), from some other store accessed by `@wordpress/data` ([example usage](https://github.com/Automattic/wp-calypso/blob/93d3b00576d8c8173cfc488d2bf097ff87165fea/client/landing/stepper/declarative-flow/internals/steps-repository/edit-email/index.tsx#L23)), or directly from the results of an API call. The `state` is not used for anything other than to fetch the purchase object.

Also, before calling this function, we are relying on the fact that the code has checked if this is a DIFM purchase. Otherwise, the output of this function does not make sense for non-DIFM purchases.

Reply:
> The calculation is on the purchase object, which could come from different places - it could be from the global redux state (like here), from some other store accessed by @wordpress/data ([example usage](https://github.com/Automattic/wp-calypso/blob/93d3b00576d8c8173cfc488d2bf097ff87165fea/client/landing/stepper/declarative-flow/internals/steps-repository/edit-email/index.tsx#L23)), or directly from the results of an API call. The state is not used for anything other than to fetch the purchase object.

Actually the intention of this piece of code was to extract the purchase details from the REDUX state slice which holds the information cached from /purchases endpoint. Nothing else. What your explaining sounds familiar to the [util method anti pattern](https://www.yanglinzhao.com/posts/utils-antipattern/) similar to util class but for functions. Or am I wrong? This function does one thing. Transforms and calculates DIFM related price tier details if available in the identified purchase of the purchases slice of the supplied redux tree and will return null if not.

> Also, before calling this function, we are relying on the fact that the code has checked if this is a DIFM purchase. Otherwise, the output of this function does not make sense for non-DIFM purchases.
Yes the code anticipates that it will be call on any variation of the universal state slice. And it will handle all those variations as appropriate :)

Reply:
I added some changes to improve the clarity of the selector 👍 
https://github.com/Automattic/wp-calypso/pull/66559/commits/7a245a208e877779cf262ea7ebf6da4ec8acb9ef

Reply:
> Transforms and calculates DIFM related price tier details if available in the identified purchase of the purchases slice of the supplied redux tree and will return null if not.

I understand that this function is currently designed to pick the purchase object from the state slice. However, IMO, this function is better suited to work on the `purchase` object itself. My reasoning:
* The function currently does not use any other part of the state to calculate the result. It only picks the finds the `purchase` object again, which we already have access to when we call this function.
* This function is executed for all types of purchases since it is being called in the `connect` function here, which seems redundant to me.
* The function is tightly coupled to the redux state, and assumes that the `purchase` object will always be in the redux state. However, it may not be the case as:
  * The [recommended approach for data is using react-query](https://github.com/Automattic/wp-calypso/blob/trunk/client/data/README.md). For future changes that involve fetching purchase data from the backend, this function will have to be refactored to work on the `purchase` object.
  * Not all data is stored in the global redux state. In the stepper framework for example, the data is stored in `@wordpress/data` stores. ([link](https://github.com/Automattic/wp-calypso/blob/trunk/packages/data-stores/README.md))

According to the article you linked to, I don't this is an anti-pattern as the `lib/purchases` directory has a clearly defined boundary (`purchase` object functions) and it is not a part of general app-wide util module.

However, let's keep this PR moving forward 🙂 If we do decide to keep the function as it is, I have added one more bit of feedback for the change in `client/me/purchases/manage-purchase/purchase-meta.jsx`.

Reply:
Oh I wasn't aware of the move to react-query TIL! . I had no idea we were evolving the data layer to something different! 

In which case trying to maintain a set of clean set of selectors is a non starter anyway (which was my main motivation)!
I will move this to the lib folder as you recommended! :)